File: python-uv-backend/scripts/verify.py
#!/usr/bin/env python3
"""
Comprehensive verification script for Python + uv projects.
Runs linting, formatting, type checking, and tests with coverage.
"""

import argparse
import subprocess
import sys
from pathlib import Path
from typing import List, Tuple


class Colors:
    """ANSI color codes for terminal output."""
    HEADER = '\033[95m'
    BLUE = '\033[94m'
    CYAN = '\033[96m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'


def print_step(message: str) -> None:
    """Print a step header."""
    print(f"\n{Colors.HEADER}{Colors.BOLD}{'='*60}{Colors.ENDC}")
    print(f"{Colors.HEADER}{Colors.BOLD}{message}{Colors.ENDC}")
    print(f"{Colors.HEADER}{Colors.BOLD}{'='*60}{Colors.ENDC}\n")


def run_command(cmd: List[str], step_name: str) -> Tuple[bool, str]:
    """Run a command and return success status and output."""
    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            check=False
        )

        output = result.stdout + result.stderr
        success = result.returncode == 0

        if success:
            print(f"{Colors.GREEN}✓ {step_name} passed{Colors.ENDC}")
        else:
            print(f"{Colors.RED}✗ {step_name} failed{Colors.ENDC}")
            if output.strip():
                print(f"{Colors.YELLOW}Output:{Colors.ENDC}")
                print(output)

        return success, output
    except FileNotFoundError:
        print(f"{Colors.RED}✗ {step_name} failed: Command not found{Colors.ENDC}")
        print(f"{Colors.YELLOW}Command: {' '.join(cmd)}{Colors.ENDC}")
        return False, "Command not found"


def check_uv_installed() -> bool:
    """Check if uv is installed."""
    success, _ = run_command(["uv", "--version"], "uv installation check")
    return success


def run_ruff_check(fix: bool = False) -> bool:
    """Run ruff linting."""
    print_step("Running Ruff Linting")
    cmd = ["uv", "run", "ruff", "check"]
    if fix:
        cmd.append("--fix")
    cmd.append(".")
    success, _ = run_command(cmd, "ruff check")
    return success


def run_ruff_format(check_only: bool = True) -> bool:
    """Run ruff formatting."""
    print_step("Running Ruff Formatting")
    cmd = ["uv", "run", "ruff", "format"]
    if check_only:
        cmd.append("--check")
    cmd.append(".")
    success, _ = run_command(cmd, "ruff format")
    return success


def run_mypy() -> bool:
    """Run mypy type checking."""
    print_step("Running MyPy Type Checking")
    success, _ = run_command(["uv", "run", "mypy", "."], "mypy")
    return success


def run_tests(coverage: bool = True, min_coverage: int = 80) -> bool:
    """Run pytest with optional coverage."""
    print_step("Running Tests with Pytest")

    cmd = ["uv", "run", "pytest"]

    if coverage:
        cmd.extend([
            "--cov=.",
            "--cov-report=term-missing",
            f"--cov-fail-under={min_coverage}"
        ])

    cmd.extend(["-v", "--tb=short"])

    success, _ = run_command(cmd, "pytest")
    return success


def main():
    parser = argparse.ArgumentParser(
        description="Verify Python + uv project (lint, format, type check, test)"
    )
    parser.add_argument(
        "--fix",
        action="store_true",
        help="Auto-fix linting issues where possible"
    )
    parser.add_argument(
        "--format",
        action="store_true",
        help="Auto-format code (not just check)"
    )
    parser.add_argument(
        "--no-coverage",
        action="store_true",
        help="Skip coverage reporting"
    )
    parser.add_argument(
        "--min-coverage",
        type=int,
        default=80,
        help="Minimum coverage percentage (default: 80)"
    )
    parser.add_argument(
        "--skip-lint",
        action="store_true",
        help="Skip linting"
    )
    parser.add_argument(
        "--skip-format",
        action="store_true",
        help="Skip format checking"
    )
    parser.add_argument(
        "--skip-mypy",
        action="store_true",
        help="Skip type checking"
    )
    parser.add_argument(
        "--skip-tests",
        action="store_true",
        help="Skip tests"
    )

    args = parser.parse_args()

    # Track overall success
    all_passed = True
    results = {}

    print(f"{Colors.CYAN}{Colors.BOLD}Python + uv Project Verification{Colors.ENDC}\n")

    # Check uv is installed
    if not check_uv_installed():
        print(f"\n{Colors.RED}ERROR: uv is not installed. Install it first:{Colors.ENDC}")
        print("  curl -LsSf https://astral.sh/uv/install.sh | sh")
        sys.exit(1)

    # Run checks
    if not args.skip_lint:
        results['lint'] = run_ruff_check(fix=args.fix)
        all_passed = all_passed and results['lint']

    if not args.skip_format:
        results['format'] = run_ruff_format(check_only=not args.format)
        all_passed = all_passed and results['format']

    if not args.skip_mypy:
        results['mypy'] = run_mypy()
        all_passed = all_passed and results['mypy']

    if not args.skip_tests:
        results['tests'] = run_tests(
            coverage=not args.no_coverage,
            min_coverage=args.min_coverage
        )
        all_passed = all_passed and results['tests']

    # Print summary
    print_step("Verification Summary")

    for check, passed in results.items():
        status = f"{Colors.GREEN}✓ PASSED{Colors.ENDC}" if passed else f"{Colors.RED}✗ FAILED{Colors.ENDC}"
        print(f"{check.ljust(15)}: {status}")

    print()

    if all_passed:
        print(f"{Colors.GREEN}{Colors.BOLD}🎉 All checks passed!{Colors.ENDC}\n")
        sys.exit(0)
    else:
        print(f"{Colors.RED}{Colors.BOLD}❌ Some checks failed{Colors.ENDC}\n")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: frontend-plugin/skills/bun-vite-react/scripts/verify.ts
#!/usr/bin/env bun
/**
 * Comprehensive verification script for Bun + Vite + React + TypeScript projects.
 * Runs linting, formatting, type checking, and tests with coverage.
 */

const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  red: '\x1b[91m',
  green: '\x1b[92m',
  yellow: '\x1b[93m',
  blue: '\x1b[94m',
  cyan: '\x1b[96m',
  magenta: '\x1b[95m',
};

interface CheckResult {
  name: string;
  passed: boolean;
  output?: string;
}

function printStep(message: string): void {
  console.log(`\n${colors.magenta}${colors.bright}${'='.repeat(60)}${colors.reset}`);
  console.log(`${colors.magenta}${colors.bright}${message}${colors.reset}`);
  console.log(`${colors.magenta}${colors.bright}${'='.repeat(60)}${colors.reset}\n`);
}

async function runCommand(
  cmd: string[],
  stepName: string
): Promise<{ success: boolean; output: string }> {
  try {
    const proc = Bun.spawn(cmd, {
      stdout: 'pipe',
      stderr: 'pipe',
    });

    const output = await new Response(proc.stdout).text();
    const error = await new Response(proc.stderr).text();
    const exitCode = await proc.exited;

    const combinedOutput = output + error;
    const success = exitCode === 0;

    if (success) {
      console.log(`${colors.green}✓ ${stepName} passed${colors.reset}`);
    } else {
      console.log(`${colors.red}✗ ${stepName} failed${colors.reset}`);
      if (combinedOutput.trim()) {
        console.log(`${colors.yellow}Output:${colors.reset}`);
        console.log(combinedOutput);
      }
    }

    return { success, output: combinedOutput };
  } catch (error) {
    console.log(`${colors.red}✗ ${stepName} failed: ${error}${colors.reset}`);
    return { success: false, output: String(error) };
  }
}

async function checkBunInstalled(): Promise<boolean> {
  const { success } = await runCommand(['bun', '--version'], 'Bun installation check');
  return success;
}

async function runESLint(fix: boolean = false): Promise<boolean> {
  printStep('Running ESLint');
  const cmd = ['bun', 'run', 'eslint', '.'];
  if (fix) {
    cmd.push('--fix');
  }
  const { success } = await runCommand(cmd, 'eslint');
  return success;
}

async function runPrettier(checkOnly: boolean = true): Promise<boolean> {
  printStep('Running Prettier');
  const cmd = ['bun', 'run', 'prettier'];
  if (checkOnly) {
    cmd.push('--check', '.');
  } else {
    cmd.push('--write', '.');
  }
  const { success } = await runCommand(cmd, 'prettier');
  return success;
}

async function runTypeCheck(): Promise<boolean> {
  printStep('Running TypeScript Type Checking');
  const { success } = await runCommand(['bun', 'run', 'tsc', '--noEmit'], 'tsc');
  return success;
}

async function runTests(coverage: boolean = true, minCoverage: number = 80): Promise<boolean> {
  printStep('Running Tests with Vitest');
  const cmd = ['bun', 'run', 'vitest', 'run'];

  if (coverage) {
    cmd.push('--coverage');
    // Note: Coverage threshold configured in vitest.config.ts
  }

  const { success } = await runCommand(cmd, 'vitest');
  return success;
}

interface CliArgs {
  fix: boolean;
  format: boolean;
  noCoverage: boolean;
  minCoverage: number;
  skipLint: boolean;
  skipFormat: boolean;
  skipTypeCheck: boolean;
  skipTests: boolean;
}

function parseArgs(): CliArgs {
  const args = process.argv.slice(2);
  return {
    fix: args.includes('--fix'),
    format: args.includes('--format'),
    noCoverage: args.includes('--no-coverage'),
    minCoverage: parseInt(args.find(a => a.startsWith('--min-coverage='))?.split('=')[1] || '80'),
    skipLint: args.includes('--skip-lint'),
    skipFormat: args.includes('--skip-format'),
    skipTypeCheck: args.includes('--skip-type-check'),
    skipTests: args.includes('--skip-tests'),
  };
}

async function main() {
  const args = parseArgs();
  const results: Record<string, boolean> = {};
  let allPassed = true;

  console.log(`${colors.cyan}${colors.bright}Bun + Vite + React + TypeScript Project Verification${colors.reset}\n`);

  // Check Bun is installed
  if (!await checkBunInstalled()) {
    console.log(`\n${colors.red}ERROR: Bun is not installed. Install it first:${colors.reset}`);
    console.log('  curl -fsSL https://bun.sh/install | bash');
    process.exit(1);
  }

  // Run checks
  if (!args.skipLint) {
    results.lint = await runESLint(args.fix);
    allPassed = allPassed && results.lint;
  }

  if (!args.skipFormat) {
    results.format = await runPrettier(!args.format);
    allPassed = allPassed && results.format;
  }

  if (!args.skipTypeCheck) {
    results.typeCheck = await runTypeCheck();
    allPassed = allPassed && results.typeCheck;
  }

  if (!args.skipTests) {
    results.tests = await runTests(!args.noCoverage, args.minCoverage);
    allPassed = allPassed && results.tests;
  }

  // Print summary
  printStep('Verification Summary');

  for (const [check, passed] of Object.entries(results)) {
    const status = passed
      ? `${colors.green}✓ PASSED${colors.reset}`
      : `${colors.red}✗ FAILED${colors.reset}`;
    console.log(`${check.padEnd(15)}: ${status}`);
  }

  console.log();

  if (allPassed) {
    console.log(`${colors.green}${colors.bright}🎉 All checks passed!${colors.reset}\n`);
    process.exit(0);
  } else {
    console.log(`${colors.red}${colors.bright}❌ Some checks failed${colors.reset}\n`);
    process.exit(1);
  }
}

main();
